### PR TITLE
[Rgen] Implement SmartEnums analyzer.

### PR DESCRIPTION
**Pull request by mandel-macaque**
Provide an analyzer that will detect the following issues with a SmartEnum binding:

 * RBI0002 (Error): Smart enum values must be tagged with an Field<EnumValue> attribute.
 * RBI0003 (Error): Smart enum backing field cannot appear more than once.
 * RBI0004 (Error): Smart enum backing field must represent a valid C# identifier to be used.
 * RBI0005 (Error): Non Apple framework bindings must provide a library name.
 * RBI0006 (Warning): Do not provide the LibraryName for known Apple frameworks.
 * RBI0007 (Error): Enum values must be tagged with Field<EnumValue>.

The analyzer has been added to the build too so that we can get compilation errors in our own SDK.

### CODE REVIEW

File: src/rgen/Microsoft.Macios.Bindings.Analyzer/BindingTypeSemanticAnalyzer.cs [hunk 1]
```diff
@@ -38,35 +35,17 @@ public override void Initialize (AnalysisContext context)
...
+	public ImmutableArray<Diagnostic> Analyze (PlatformName _, ClassDeclarationSyntax declarationNode, INamedTypeSymbol symbol)
+	{
+		if (declarationNode.Modifiers.Any (x => x.IsKind (SyntaxKind.PartialKeyword)))
+			return [];
 
-		var boundAttributes = classSymbol.GetAttributes ();
-		if (boundAttributes.Length == 0) {
-			return;
-		}
+		var diagnostic = Diagnostic.Create (RBI0001,
```
Comment (rolfbjarne) [inline src/rgen/Microsoft.Macios.Bindings.Analyzer/BindingTypeSemanticAnalyzer.cs:45]: Add the actual error message as a comment:

```suggestion
		var diagnostic = Diagnostic.Create (RBI0001, // Binding types should be declared as partial classes.
```

File: src/rgen/Microsoft.Macios.Bindings.Analyzer/Resources.resx [hunk 2]
```diff
@@ -34,4 +30,82 @@
         <value>Binding type declaration must be partial</value>
         <comment>The title of the diagnostic.</comment>
     </data>
+
+    <data name="RBI0002Description" xml:space="preserve">
+        <value>In order for the code to be generated a smart enum value has to have a backing field.</value>
+        <comment>An optional longer localizable description of the diagnostic.</comment>
```
Comment (rolfbjarne) [inline src/rgen/Microsoft.Macios.Bindings.Analyzer/Resources.resx:36]: Do we need the repeated comment nodes for each error message?

File: src/rgen/Microsoft.Macios.Bindings.Analyzer/SmartEnumsAnalyzer.cs [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+		var members = symbol.GetMembers ().OfType<IFieldSymbol> ().ToArray ();
+		// we do not allow backing fields to be duplicated, keep track of those added so far to raise
+		// a diagnostic if we find a duplicate
+		var backingFields = new HashSet<string> ();
+		foreach (var fieldSymbol in members) {
+			var attributes = fieldSymbol.GetAttributeData ();
+			if (attributes.Count == 0) {
+				//  All enum values are marked with a Field attribute, therefore add a diagnostic
+				bucket.Add (Diagnostic.Create (
+					RBI0002, // Smart enum values must be tagged with an Field<EnumValue> attribute.
```
Comment (rolfbjarne) [inline src/rgen/Microsoft.Macios.Bindings.Analyzer/SmartEnumsAnalyzer.cs:163]: 👍

File: tests/rgen/Microsoft.Macios.Bindings.Analyzer.Tests/SmartEnumsAnalyzerTests.cs [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+	public async Task SmartEnumAppleFrameworkNotLibrary (ApplePlatform platform, string inputText)
+	{
+		var (compilation, _) = CreateCompilation (nameof (SmartEnumsAnalyzerTests), platform, inputText);
+		var diagnostics = await RunAnalyzer (new SmartEnumsAnalyzer (), compilation);
+
+		var analyzerDiagnotics = diagnostics
+			.Where (d => d.Id == SmartEnumsAnalyzer.RBI0006.Id).ToArray ();
+		Assert.Single (analyzerDiagnotics);
+		VerifyDiagnosticMessage (analyzerDiagnotics [0], SmartEnumsAnalyzer.RBI0006.Id,
+			DiagnosticSeverity.Warning, "The backing Field of 'AVFoundation.AVCaptureSystemPressureExampleLevel.Shutdown' should not provide a LibraryName");
```
Comment (rolfbjarne) [inline tests/rgen/Microsoft.Macios.Bindings.Analyzer.Tests/SmartEnumsAnalyzerTests.cs:199]: I think a message like this sounds a bit better:

> The Field attribute for the enum value 'AVFoundation.AVCaptureSystemPressureExampleLevel.Shutdown' must not provide a value for LibraryName

File: tests/rgen/Microsoft.Macios.Bindings.Analyzer.Tests/SmartEnumsAnalyzerTests.cs [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+	[AllSupportedPlatforms (customLibraryEmptyLibraryNameParameter)]
+	public async Task SmartEnumThirdPartyLibrary (ApplePlatform platform, string inputText)
+	{
+		var (compilation, _) = CreateCompilation (nameof (SmartEnumsAnalyzerTests), platform, inputText);
+		var diagnostics = await RunAnalyzer (new SmartEnumsAnalyzer (), compilation);
+		var analyzerDiagnotics = diagnostics
+			.Where (d => d.Id == SmartEnumsAnalyzer.RBI0005.Id).ToArray ();
+		Assert.Single (analyzerDiagnotics);
+		VerifyDiagnosticMessage (analyzerDiagnotics [0], SmartEnumsAnalyzer.RBI0005.Id,
+			DiagnosticSeverity.Error, "The enum value 'CustomLibrary.CustomLibraryEnum.High' backing field must provide a library name");
```
Comment (rolfbjarne) [inline tests/rgen/Microsoft.Macios.Bindings.Analyzer.Tests/SmartEnumsAnalyzerTests.cs:276]: > The field attribute for the enum value '...' must set the property 'LibraryName'

File: tests/rgen/Microsoft.Macios.Bindings.Analyzer.Tests/SmartEnumsAnalyzerTests.cs [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+	Shutdown,
+}";
+
+		var (compilation, _) = CreateCompilation (nameof (SmartEnumsAnalyzerTests), platform, inputText);
+		var diagnostics = await RunAnalyzer (new SmartEnumsAnalyzer (), compilation);
+		var analyzerDiagnotics = diagnostics
+			.Where (d => d.Id == SmartEnumsAnalyzer.RBI0003.Id).ToArray ();
+		Assert.Single (analyzerDiagnotics);
+		VerifyDiagnosticMessage (analyzerDiagnotics [0], SmartEnumsAnalyzer.RBI0003.Id,
+			DiagnosticSeverity.Error, "The enum value 'AVFoundation.AVCaptureSystemPressureExampleLevel.Fair' backing field 'AVCaptureSystemPressureLevelNominal' is already in use");
```
Comment (rolfbjarne) [inline tests/rgen/Microsoft.Macios.Bindings.Analyzer.Tests/SmartEnumsAnalyzerTests.cs:344]: > The backing field '...' for the enum value '...' is already in use.

Also, can we provide a diagnostic for the first definition too, or alternatively something like:

> The backing field '...' for the enum value '...' is already in use for the enum value '...'

File: src/rgen/Microsoft.Macios.Bindings.Analyzer/SmartEnumsAnalyzer.cs [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+				var fieldAttrData = fieldAttrDataList [0];
+				var fieldSyntax = fieldAttrData.ApplicationSyntaxReference?.GetSyntax ();
+				if (fieldSyntax is null) {
+					continue;
+				}
+
+				if (FieldData<EnumValue>.TryParse (fieldAttrData, out var fieldData, out var errorTuple)) {
+					// only provide diagnostics if we managed to parse the FieldData, else we have a bug in the
+					// analyzer
+					if (!backingFields.Add (fieldData.Value.SymbolName)) {
```
Comment (stephen-hawley) [inline src/rgen/Microsoft.Macios.Bindings.Analyzer/SmartEnumsAnalyzer.cs:181]: I like this test

File: src/rgen/Microsoft.Macios.Bindings.Analyzer/SmartEnumsAnalyzer.cs [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+						// All symbol names have to be unique
+						bucket.Add (Diagnostic.Create (
+							RBI0003, // Smart enum backing field cannot appear more than once.
+							fieldSyntax.GetLocation (),
+							fieldSymbol.ToDisplayString (), fieldData.Value.SymbolName));
+					}
+
+					if (!isAppleFramework) {
+						// If the Field attribute is not from a known apple library, the library name is set
+						if (string.IsNullOrWhiteSpace (fieldData.Value.LibraryName)) {
```
Comment (stephen-hawley) [inline src/rgen/Microsoft.Macios.Bindings.Analyzer/SmartEnumsAnalyzer.cs:191]: This test belies your error message. You probably want to change your error message to be "null or empty" or break out your cases into:
empty, obviously incorrect path, likely correct path. And if you go this route, I'd break it out into a separate function:
```csharp
bool TryGetErrorForLibrary (bool isAppleFrameWork, string? libraryName, [NotNullWhen(true)] out Diagnostic? error)
{
     // implementation here
}
// then your call site here is:
if (TryGetErrorForLibrary (isAppleFrameWork, fieldData.Value.LibraryName, out var libraryError))
    bucket.Add (libraryError);
```

**Review by stephen-hawley — CHANGES_REQUESTED** [review]

File: tests/rgen/Microsoft.Macios.Bindings.Analyzer.Tests/SmartEnumsAnalyzerTests.cs (hunk 6 above)
Comment (mandel-macaque) [reply tests/rgen/Microsoft.Macios.Bindings.Analyzer.Tests/SmartEnumsAnalyzerTests.cs:344]: I can do that, move from a hashset to a dict and update the message.

**Review by dalexsoto — APPROVED** [review]

**Pull request merged**